### ISSUE TITLE
[HOLD for payment 2024-04-09] [HOLD for payment 2024-03-13] [$500] Video - Play&Pause buttons don't function in preview mode if expanded the video before

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#36873]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.43-0
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:**

## Action Performed:
Prerequisites: The user should be logged in and a video exist in the chat
1. Go to prerequisite chat
2. Play&Pause the video
3. Tap on the expand button on the video preview
4. Play&Pause the video
5. Exit from the attachment view via clicking X button top right
6. Try to Play&Pause the video in preview mode

## Expected Result:
Play&Pause buttons function correctly

## Actual Result:
Play&Pause buttons don't function in preview mode and console error occurs

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/ae83af82-c96f-44c4-b93b-2e182fd878d2

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~013093a3e364e94fed</li>
        <li>Upwork Job ID: 1759909332125061120</li>
        <li>Last Price Increase: 2024-02-20</li>
    </ul>
</details>

**Comment by izarutskaya** [comment]
We think that this bug might be related to #vip-vsb
CC @quinthar

**Comment by lakchote** [comment]
This looks related to https://github.com/Expensify/App/pull/30829

cc @francoisl @Skalakid @akinwale

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The play/pause button doesn't work after expanding and closing the video modal.

### What is the root cause of that problem?
From the video, we can see there is a console error. The error indicates that the video player is not ready yet and it's because the video player element/ref doesn't exist anymore. 

We have a `useEffect` here that is responsible for setting the video player ref.
https://github.com/Expensify/App/blob/7430c5f93df630ed20c80caffcd5a7f05289c98c/src/components/VideoPlayer/BaseVideoPlayer.js#L139-L143

Let's say we play a video A from the video preview. The video player ref is first set to the video preview A. Then, we press the expand button to see the video in the attachment modal. Currently, the video player ref is updated to the video player in the attachment modal (this is wrong). After we close the attachment modal, the video player doesn't exist anymore, so it always throws an error.

However, the expected result is, when we expand the video to view it in the attachment modal, the ref shouldn't be updated because we early return if `shouldUseSharedVideoElement` is true.
https://github.com/Expensify/App/blob/7430c5f93df630ed20c80caffcd5a7f05289c98c/src/components/VideoPlayer/BaseVideoPlayer.js#L139-L143

And `shouldUseSharedVideoElement` is true if it's used in the carousel.
https://github.com/Expensify/App/blob/7430c5f93df630ed20c80caffcd5a7f05289c98c/src/components/Attachments/AttachmentView/index.js#L208

`shouldUseSharedVideoElement` basically shares the video player between components. So, the video player state between video preview and video in the carousel.

However, `isUsedInCarousel` is always false because we never pass it.
### What changes do you think we should make in order to solve the problem?

Pass `isUsedInCarousel` from `CarouselItem`.
https://github.com/Expensify/App/blob/7430c5f93df630ed20c80caffcd5a7f05289c98c/src/components/Attachments/AttachmentCarousel/CarouselItem.js#L102-L111

**Comment by DylanDylann** [comment]
The solution for this issue is the same [this one](https://github.com/Expensify/App/issues/36634#issuecomment-1952932806) and @tienifr proposed this issue before. I think we should handle this issue in https://github.com/Expensify/App/issues/36634

**Comment by DylanDylann** [comment]
@lakchote @Puneet-here This deploy blocker has the same solution with [another issue](https://github.com/Expensify/App/issues/36634) that was created before, @tienifr proposed the correct proposal first and I also reviewed it before this deploy blocker is created. Now both these issues are waiting for an official assignment from the internal engineer. I think we should close one of them and move forward with the remaining one.

**Comment by puneetlath** [comment]
Maybe let's just have whoever can raise the PR fastest raise it. @tienifr @bernhardoj are either of you able to do that now? And we can share the bounty between everyone.

**Comment by DylanDylann** [comment]
@puneetlath Considering this issue has same RCA and solution as [an earlier issue](https://github.com/Expensify/App/issues/36634), I think this issue should be considered a dupe and closed. We can raise the PR on the original issue which already has the assignment: https://github.com/Expensify/App/issues/36634

**Comment by puneetlath** [comment]
Ok closing as we're handling in https://github.com/Expensify/App/issues/36634.

**Comment by Skalakid** [comment]
@DylanDylann @puneetlath the problem is more complex and probably has a different root cause than https://github.com/Expensify/App/issues/36634. During video player feature development we had the same problem with Play&Pause buttons as in this issue. It was fixed and it works properly on the video-player branch. Probably some other PRs or merging with the main broke it. 

The root of the problem is that whole video sharing functionality doesn't work now. And after exiting modal strange things may appear with the player. I think the fix will be similar as before so I will create PR.

**Comment by DylanDylann** [comment]
@Skalakid  Yes, I agree with you. Because @ishpaul777 [approved](https://github.com/Expensify/App/issues/36873#issuecomment-1954421662) the proposal, at that time I didn't test the solution carefully for this issue. While implementing PR, we see that this issue is more complicated (more detail [here](https://github.com/Expensify/App/pull/36921#issuecomment-1956080552)). @puneetlath I also think we should fix this issue separately to avoid blocking [this PR](https://github.com/Expensify/App/pull/36921)

**Comment by puneetlath** [comment]
Ah ok, shall we reopen this issue then @Skalakid?

**Comment by Skalakid** [comment]
@puneetlath Oh sorry, I think is shouldn't be a deploy blocker. I don't know why melvin marked it as blocker after I commented about this issue 😅 Also I think we can close this issue since the [fix](https://github.com/Expensify/App/pull/37202) was merged :D

**Comment by Skalakid** [comment]
@ishpaul777 Yes, I'm investigating the issue right now

**Comment by Skalakid** [comment]
@ishpaul777 @aimane-chnaif I created a follow-up draft PR with the fix for Play&Pause button that enables video sharing feature once again. On Monday I add last changes and test my changes on all platforms so I can give it to review.

The main problem that will be fixed in m PR is the fact that when entering full-screen mode, the whole app rerenders due to `useWindowDimensions()` hook that checks `windowWidth` and `windowHeight`. Because of that many bugs appear. In my PR I've added dimension locking, so every time user enters fullscreen mode app dimensions won't change.
Because of that, we will be able to use the shared video feature and also handle screen rotation on all native devices

**Comment by Skalakid** [comment]
@ishpaul777 I opened the PR for review

**Comment by lakchote** [comment]
> i dont think the regression was from the initial PR directly, the RC was much complex than initially expected, also the only a small part was reverted from the first PR so i feel there shouldn't a penalty in this case, @lakchote can you please help with evaluation here as assigned engineer

Even though the RC was complex to fix, this comes directly from the initial PR related to the video player and files concerned (`BaseVideoPlayer` for instance). This case should have been identified and fixed from the beginning, even from just a functional standpoint.

That's why, since it's a regression, I think 50% payment seems fair to me.

**Comment by miljakljajic** [comment]
Paid 250 via upwork to @ishpaul777 - thank you all
